engine/engine.go

```diff
@@ -0,0 +1,352 @@
+package engine
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"strings"
+	"sync"
+
+	"github.com/hashicorp/go-hclog"
+
+	"google.golang.org/grpc/credentials/insecure"
+
+	"github.com/gruntwork-io/go-commons/errors"
+	"github.com/gruntwork-io/terragrunt-engine-go/engine"
+	"github.com/gruntwork-io/terragrunt-engine-go/proto"
+	"github.com/gruntwork-io/terragrunt/options"
+	"github.com/gruntwork-io/terragrunt/util"
+	"github.com/hashicorp/go-plugin"
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/types/known/anypb"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+const (
+	engineVersion     = 1
+	engineCookieKey   = "engine"
+	engineCookieValue = "terragrunt"
+)
+
+var (
+	engineClients = sync.Map{}
+)
+
+type ExecutionOptions struct {
+	TerragruntOptions *options.TerragruntOptions
+	CmdStdout         io.Writer
+	CmdStderr         io.Writer
+	WorkingDir        string
+	SuppressStdout    bool
+	AllocatePseudoTty bool
+	Command           string
+	Args              []string
+}
+
+type engineInstance struct {
+	terragruntEngine *proto.EngineClient
+	client           *plugin.Client
+	executionOptions *ExecutionOptions
+}
+
+// IsEngineEnabled returns true if the experimental engine is enabled.
+func IsEngineEnabled() bool {
+	switch strings.ToLower(os.Getenv("TG_EXPERIMENTAL_ENGINE")) {
+	case "1", "yes", "true", "on":
+		return true
+	}
+	return false
+}
+
+// Run executes the given command with the experimental engine.
+func Run(
+	ctx context.Context,
+	runOptions *ExecutionOptions,
+) (*util.CmdOutput, error) {
+	workingDir := runOptions.TerragruntOptions.WorkingDir
+	instance, found := engineClients.Load(workingDir)
+	// initialize engine for working directory
+	if !found {
+		terragruntEngine, client, err := createEngine(runOptions.TerragruntOptions)
+		if err != nil {
+			return nil, errors.WithStackTrace(err)
+		}
+		engineClients.Store(workingDir, &engineInstance{
+			terragruntEngine: terragruntEngine,
+			client:           client,
+			executionOptions: runOptions,
+		})
+		instance, _ = engineClients.Load(workingDir)
+		if err := initialize(ctx, runOptions, terragruntEngine); err != nil {
+			return nil, errors.WithStackTrace(err)
+		}
+	}
+
+	terragruntEngine := instance.(*engineInstance).terragruntEngine
```

Comment:
```suggestion
        engineInstance, ok := instance.(*engineInstance); 
        if !ok {
	  return errors.Erroff(".....")
	}
```

I would add a trivial type check